packages/babel-traverse/src/path/conversion.js

```diff
@@ -223,10 +223,20 @@ function hoistFunctionEnvironment(
     );
 
     argumentsPaths.forEach(argumentsChild => {
-      const argsRef = t.identifier(argumentsBinding);
-      argsRef.loc = argumentsChild.node.loc;
-
-      argumentsChild.replaceWith(argsRef);
+      let currentScope = argumentsChild.scope;
+      do {
+        if (
+          !currentScope.hasOwnBinding("arguments") &&
+          !(
+            currentScope.path.isFunction() &&
+            !currentScope.path.isArrowFunctionExpression()
```

Comment:
I think these checks should be moved to `getScopeInformation`, where we push to the `argumentsPaths` array.

Also, I find this logic with nested negations a bit hard to read. Maybe (when moving it to `getScopeInformation`) we could rewrite it to use an early return?

```js
let curr = path.scope;
do {
  if (curr.hasOwnBinding("arguments")) return;
  if (curr.path.isFunction() && !currentScope.path.isArrowFunctionExpression()) break;
} while ((curr = curr.parent));
```